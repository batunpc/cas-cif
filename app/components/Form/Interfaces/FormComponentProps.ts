import { FormProps } from "@rjsf/core";

type FormComponentProps<T = any> = Partial<FormProps<T>>;

export default FormComponentProps;
