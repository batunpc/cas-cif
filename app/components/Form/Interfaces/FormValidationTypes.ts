export interface ISupportExternalValidation {
  selfValidate(): any[];
}

export interface ValidatingFormProps {
  setValidatingForm: (validatingForm: ISupportExternalValidation) => void;
}
