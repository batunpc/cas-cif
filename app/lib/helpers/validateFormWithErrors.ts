export default function validateFormWithErrors(formObject: HTMLFormElement) {
  formObject.onSubmit({
    preventDefault: () => {},
    persist: () => {},
  });
  // Effectively validating the form a second time to retrieve the errors
  const validationResult = formObject.validate(formObject.state.formData);

  return validationResult.errors;
}
