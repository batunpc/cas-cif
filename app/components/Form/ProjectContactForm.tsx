import { JSONSchema7, JSONSchema7Definition } from "json-schema";
import EmptyObjectFieldTemplate from "lib/theme/EmptyObjectFieldTemplate";
import { useMemo, useRef } from "react";
import { graphql, useFragment, useMutation } from "react-relay";
import { ProjectContactForm_query$key } from "__generated__/ProjectContactForm_query.graphql";
import FormBase from "./FormBase";
import Grid from "@button-inc/bcgov-theme/Grid";
import FormBorder from "lib/theme/components/FormBorder";
import { Button } from "@button-inc/bcgov-theme";
import { mutation as addContactToRevisionMutation } from "mutations/Contact/addContactToRevision";
import { mutation as deleteFormChangeMutation } from "mutations/FormChange/deleteFormChange";
import { mutation as updateFormChangeMutation } from "mutations/FormChange/updateFormChange";
import projectContactSchema from "data/jsonSchemaForm/projectContactSchema";
import useDebouncedMutation from "mutations/useDebouncedMutation";
import { ConnectionHandler } from "react-relay";
import { ValidatingFormProps } from "./Interfaces/FormValidationTypes";
import validateFormWithErrors from "lib/helpers/validateFormWithErrors";

interface Props extends ValidatingFormProps {
  query: ProjectContactForm_query$key;
}

const uiSchema = {
  contactId: {
    "ui:placeholder": "Select a Contact",
    "ui:col-md": 12,
    "bcgov:size": "small",
    "ui:widget": "SearchWidget",
    "ui:options": {
      label: false,
    },
  },
};

const ProjectContactForm: React.FC<Props> = (props) => {
  const formRefs = useRef({});

  const { query } = useFragment(
    graphql`
      fragment ProjectContactForm_query on Query {
        query {
          projectRevision(id: $projectRevision) {
            id
            rowId
            formChangesByProjectRevisionId(
              filter: { formDataTableName: { equalTo: "project_contact" } }
              first: 2147483647
            ) @connection(key: "connection_formChangesByProjectRevisionId") {
              __id
              edges {
                node {
                  id
                  newFormData
                }
              }
            }
          }
          allContacts {
            edges {
              node {
                rowId
                fullName
              }
            }
          }
        }
      }
    `,
    props.query
  );

  const contactSchema = useMemo(() => {
    const schema = projectContactSchema;
    schema.properties.contactId.anyOf = query.allContacts.edges.map(
      ({ node }) => {
        return {
          type: "number",
          title: node.fullName,
          enum: [node.rowId],
          value: node.rowId,
        } as JSONSchema7Definition;
      }
    );
    return schema as JSONSchema7;
  }, [query]);

  const [addContactMutation] = useMutation(addContactToRevisionMutation);

  const addContact = (contactIndex: number) => {
    addContactMutation({
      variables: {
        input: {
          revisionId: query.projectRevision.rowId,
          contactIndex: contactIndex,
        },
        connections: [
          query.projectRevision.formChangesByProjectRevisionId.__id,
        ],
      },
    });
  };

  const [discardFormChange] = useMutation(deleteFormChangeMutation);
  const deleteContact = (formChangeId: string) => {
    const date = new Date().toISOString();
    discardFormChange({
      variables: {
        input: {
          id: formChangeId,
          formChangePatch: {
            deletedAt: date,
          },
        },
      },
      optimisticResponse: {
        updateFormChange: {
          __typename: "UpdateFormChangePayload",
        },
      },
      updater: (store) => {
        const connectionId = ConnectionHandler.getConnectionID(
          query.projectRevision.id,
          "connection_formChangesByProjectRevisionId"
        );
        const connectionRecord = store.get(connectionId);
        ConnectionHandler.deleteNode(connectionRecord, formChangeId);
      },
      onCompleted: () => {
        delete formRefs.current[formChangeId];
      },
    });
  };

  const [applyUpdateFormChangeMutation] = useDebouncedMutation(
    updateFormChangeMutation
  );
  const updateFormChange = (formChangeId: string, formData: any) => {
    applyUpdateFormChangeMutation({
      variables: {
        input: {
          id: formChangeId,
          formChangePatch: {
            newFormData: formData,
          },
        },
      },
      optimisticResponse: {
        updateFormChange: {
          formChange: {
            id: formChangeId,
            newFormData: formData,
          },
        },
      },
      debounceKey: formChangeId,
    });
  };

  const allForms = useMemo(() => {
    const contactForms = [
      ...query.projectRevision.formChangesByProjectRevisionId.edges.map(
        ({ node }) => node
      ),
    ];
    contactForms.sort(
      (a, b) => a.newFormData.contactIndex - b.newFormData.contactIndex
    );
    return contactForms;
  }, [query]);

  const [primaryContactForm, ...alternateContactForms] = allForms;

  props.setValidatingForm({
    selfValidate: () => {
      return Object.keys(formRefs.current).reduce((agg, formId) => {
        const formObject = formRefs.current[formId];
        return [...agg, ...validateFormWithErrors(formObject)];
      }, []);
    },
  });

  const clearPrimaryContact = () => {
    const { contactId, ...newFormData } = primaryContactForm.newFormData;
    updateFormChange(primaryContactForm.id, newFormData);
  };

  return (
    <>
      <Grid cols={10} align="center">
        <Grid.Row>
          <Grid.Col span={10}>
            <FormBorder title="Project Contacts">
              <Grid.Row>
                <Grid.Col span={10} className="right-aligned-column">
                  <Button>View Contact List</Button>
                </Grid.Col>
              </Grid.Row>
              <Grid.Row>
                <label>Primary Contact</label>
              </Grid.Row>
              <Grid.Row>
                <Grid.Col span={6}>
                  <FormBase
                    id="primaryContactForm"
                    ref={(el) => (formRefs.current[primaryContactForm.id] = el)}
                    formData={primaryContactForm.newFormData}
                    onChange={(change) => {
                      updateFormChange(primaryContactForm.id, change);
                    }}
                    schema={contactSchema}
                    uiSchema={uiSchema}
                    ObjectFieldTemplate={EmptyObjectFieldTemplate}
                  />
                </Grid.Col>
                <Grid.Col span={4}>
                  <Button variant="secondary" size="small">
                    Show Details
                  </Button>
                  <Button
                    variant="secondary"
                    size="small"
                    onClick={clearPrimaryContact}
                  >
                    Clear
                  </Button>
                </Grid.Col>
              </Grid.Row>
              <Grid.Row>
                <label>Secondary Contacts</label>
              </Grid.Row>
              {alternateContactForms.map((form) => (
                <Grid.Row key={form.id}>
                  <Grid.Col span={6}>
                    <FormBase
                      id={`form-${form.id}`}
                      ref={(el) => (formRefs.current[form.id] = el)}
                      formData={form.newFormData}
                      onChange={(change) => {
                        updateFormChange(form.id, change);
                      }}
                      schema={contactSchema}
                      uiSchema={uiSchema}
                      ObjectFieldTemplate={EmptyObjectFieldTemplate}
                    />
                  </Grid.Col>
                  <Grid.Col span={4}>
                    <Button
                      variant="secondary"
                      size="small"
                      onClick={() => deleteContact(form.id)}
                    >
                      Remove
                    </Button>
                  </Grid.Col>
                </Grid.Row>
              ))}

              <Grid.Row>
                <Grid.Col span={10}>
                  <Button
                    style={{ marginRight: "auto" }}
                    onClick={() =>
                      // allForms is already sorted by contactIndex
                      addContact(
                        allForms[allForms.length - 1].newFormData.contactIndex +
                          1
                      )
                    }
                  >
                    Add
                  </Button>
                </Grid.Col>
              </Grid.Row>
            </FormBorder>
          </Grid.Col>
        </Grid.Row>
      </Grid>
      <style jsx>{`
        :global(button.pg-button) {
          margin-left: 0.4em;
          margin-right: 0em;
        }
        :global(.right-aligned-column) {
          display: flex;
          justify-content: flex-end;
          align-items: flex-start;
        }
      `}</style>
    </>
  );
};

export default ProjectContactForm;
