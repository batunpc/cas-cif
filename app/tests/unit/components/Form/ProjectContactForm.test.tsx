import { render, screen } from "@testing-library/react";
import { ValidatingFormProps } from "components/Form/Interfaces/FormValidationTypes";
import ProjectContactForm from "components/Form/ProjectContactForm";
import {
  graphql,
  RelayEnvironmentProvider,
  useLazyLoadQuery,
} from "react-relay";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import compiledProjectContactFormQuery, {
  ProjectContactFormQuery,
} from "__generated__/ProjectContactFormQuery.graphql";
import validateFormWithErrors from "lib/helpers/validateFormWithErrors";
import { mocked } from "jest-mock";

jest.mock("lib/helpers/validateFormWithErrors");

const loadedQuery = graphql`
  query ProjectContactFormQuery($projectRevision: ID!) @relay_test_operation {
    query {
      # Spread the fragment you want to test here
      ...ProjectContactForm_query
    }
  }
`;

const props: ValidatingFormProps = {
  setValidatingForm: jest.fn(),
};

let environment;
const TestRenderer = () => {
  const data = useLazyLoadQuery<ProjectContactFormQuery>(loadedQuery, {
    projectRevision: "test-project-revision",
  });
  return <ProjectContactForm {...props} query={data.query} />;
};
const renderProjectForm = () => {
  return render(
    <RelayEnvironmentProvider environment={environment}>
      <TestRenderer />
    </RelayEnvironmentProvider>
  );
};

const getMockQueryPayload = () => ({
  Query() {
    return {
      projectRevision: {
        id: "Test Project Revision ID",
        rowId: 1234,
        formChangesByProjectRevisionId: {
          edges: [
            {
              node: {
                id: "Form ID 1",
                newFormData: {
                  projectId: 10,
                  contactId: 2,
                  contactIndex: 1,
                },
              },
            },
            {
              node: {
                id: "Form ID 2",
                newFormData: {
                  projectId: 10,
                  contactId: 3,
                  contactIndex: 2,
                },
              },
            },
            {
              node: {
                id: "Form ID 3",
                newFormData: {
                  projectId: 10,
                  contactId: 1,
                  contactIndex: 5,
                },
              },
            },
          ],
        },
      },
      allContacts: {
        edges: [
          {
            node: {
              rowId: 1,
              fullName: "Mister Test",
            },
          },
          {
            node: {
              rowId: 2,
              fullName: "Mister Test 2",
            },
          },
          {
            node: {
              rowId: 3,
              fullName: "Mister Test 3",
            },
          },
        ],
      },
    };
  },
});

describe("The ProjectContactForm", () => {
  beforeEach(() => {
    jest.resetAllMocks();

    environment = createMockEnvironment();

    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, getMockQueryPayload())
    );

    environment.mock.queuePendingOperation(compiledProjectContactFormQuery, {});
  });

  it("Matches the snapshot with a primary contact and multiple secondary contacts", () => {
    const componentUnderTest = renderProjectForm();
    expect(componentUnderTest.container).toMatchSnapshot();

    expect(screen.getAllByRole("textbox")).toHaveLength(3);

    // Remove buttons only appear on alternate contacs
    expect(screen.getAllByText("Remove")).toHaveLength(2);
  });

  it("Calls the addContactToRevision mutation when the Add button is clicked", () => {
    const mutationSpy = jest.fn();
    jest
      .spyOn(require("react-relay"), "useMutation")
      .mockImplementation(() => [mutationSpy, jest.fn()]);

    renderProjectForm();
    const addButton = screen.getByText("Add");
    addButton.click();

    expect(mutationSpy).toHaveBeenCalledWith({
      variables: {
        connections: expect.any(Array),
        input: {
          revisionId: 1234,
          contactIndex: 6,
        },
      },
    });
  });

  it("Calls the updateFormChange mutation when the remove button is clicked", () => {
    const mutationSpy = jest.fn();
    jest
      .spyOn(require("react-relay"), "useMutation")
      .mockImplementation(() => [mutationSpy, jest.fn()]);

    renderProjectForm();
    const removeButton = screen.getAllByText("Remove")[0];
    removeButton.click();

    expect(mutationSpy).toHaveBeenCalledWith({
      variables: {
        input: {
          id: "Form ID 2",
          formChangePatch: {
            deletedAt: expect.any(String),
          },
        },
      },
      onCompleted: expect.any(Function),
      updater: expect.any(Function),
      optimisticResponse: {
        updateFormChange: {
          __typename: "UpdateFormChangePayload",
        },
      },
    });
  });

  it("Validates all contact forms when validator is called", () => {
    mocked(validateFormWithErrors).mockImplementation(() => []);
    jest
      .spyOn(require("react-relay"), "useMutation")
      .mockImplementation(() => [jest.fn(), jest.fn()]);

    renderProjectForm();

    expect(props.setValidatingForm).toHaveBeenCalledWith({
      selfValidate: expect.any(Function),
    });

    props.setValidatingForm.mock.calls[0][0].selfValidate();

    // Once per form
    expect(mocked(validateFormWithErrors)).toHaveBeenCalledTimes(3);
  });
});
