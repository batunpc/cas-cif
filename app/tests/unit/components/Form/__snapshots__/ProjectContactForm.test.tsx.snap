// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProjectContactForm Matches the snapshot with a primary contact and multiple secondary contacts 1`] = `
<div>
  <div
    class="sc-jRQBWg cDwDrH pg-grid-container"
    cols="10"
  >
    <div
      class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
    >
      <div
        class="sc-furwcr sc-pVTFL XpgcI eqAbko pg-grid-col"
        width="100"
      >
        <fieldset
          class="jsx-4178779636"
        >
          <legend
            class="jsx-4178779636"
          >
            Project Contacts
          </legend>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <div
              class="sc-furwcr sc-pVTFL XpgcI eqAbko pg-grid-col right-aligned-column"
              width="100"
            >
              <button
                class="sc-jcFjpl dFodJy pg-button"
                id="rfj9pgbamipnij312z1gfh"
              >
                View Contact List
              </button>
            </div>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <label
              class="jsx-1069952119"
            >
              Primary Contact
            </label>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <div
              class="sc-furwcr sc-pVTFL irDrcf eqAbko pg-grid-col"
              width="60"
            >
              <form
                class="rjsf"
                id="primaryContactForm"
                novalidate=""
              >
                <div
                  class="jsx-3435786090"
                >
                  <div
                    class="jsx-3435786090"
                  >
                    <div
                      aria-expanded="false"
                      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-symcig-MuiAutocomplete-root"
                      role="combobox"
                    >
                      <div
                        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-invalid="false"
                            autocapitalize="none"
                            autocomplete="off"
                            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                            id="search-dropdown-root_contactId"
                            placeholder="Select a Contact"
                            spellcheck="false"
                            type="text"
                            value="Mister Test 2"
                          />
                          <div
                            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                          >
                            <button
                              aria-label="Clear"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                              tabindex="-1"
                              title="Clear"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                                data-testid="CloseIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                            <button
                              aria-label="Open"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                              tabindex="-1"
                              title="Open"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                                data-testid="ArrowDropDownIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M7 10l5 5 5-5z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                          </div>
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-nnbavb"
                            >
                              <span
                                class="notranslate"
                              >
                                â€‹
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="sc-furwcr sc-pVTFL bXKSoN eqAbko pg-grid-col right-aligned-column"
              width="40"
            >
              <button
                class="sc-jcFjpl VmKuo pg-button"
                id="9ypccs8jjzu68oe6ez5v6w"
              >
                Show Details
              </button>
              <button
                class="sc-jcFjpl VmKuo pg-button"
                id="xnqqccmwgulwah4wxcp7e"
              >
                Clear
              </button>
            </div>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <label
              class="jsx-1069952119"
            >
              Secondary Contacts
            </label>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <div
              class="sc-furwcr sc-pVTFL kYaPVx eqAbko pg-grid-col"
              width="80"
            >
              <form
                class="rjsf"
                id="form-Form ID 2"
                novalidate=""
              >
                <div
                  class="jsx-3435786090"
                >
                  <div
                    class="jsx-3435786090"
                  >
                    <div
                      aria-expanded="false"
                      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-symcig-MuiAutocomplete-root"
                      role="combobox"
                    >
                      <div
                        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-invalid="false"
                            autocapitalize="none"
                            autocomplete="off"
                            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                            id="search-dropdown-root_contactId"
                            placeholder="Select a Contact"
                            spellcheck="false"
                            type="text"
                            value="Mister Test 3"
                          />
                          <div
                            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                          >
                            <button
                              aria-label="Clear"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                              tabindex="-1"
                              title="Clear"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                                data-testid="CloseIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                            <button
                              aria-label="Open"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                              tabindex="-1"
                              title="Open"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                                data-testid="ArrowDropDownIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M7 10l5 5 5-5z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                          </div>
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-nnbavb"
                            >
                              <span
                                class="notranslate"
                              >
                                â€‹
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="sc-furwcr sc-pVTFL kGJjgX eqAbko pg-grid-col right-aligned-column"
              width="20"
            >
              <button
                class="sc-jcFjpl VmKuo pg-button"
                id="cny8ih6ca8t5ewg2nr6dsp"
              >
                Remove
              </button>
            </div>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <div
              class="sc-furwcr sc-pVTFL kYaPVx eqAbko pg-grid-col"
              width="80"
            >
              <form
                class="rjsf"
                id="form-Form ID 3"
                novalidate=""
              >
                <div
                  class="jsx-3435786090"
                >
                  <div
                    class="jsx-3435786090"
                  >
                    <div
                      aria-expanded="false"
                      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-symcig-MuiAutocomplete-root"
                      role="combobox"
                    >
                      <div
                        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                      >
                        <div
                          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-invalid="false"
                            autocapitalize="none"
                            autocomplete="off"
                            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                            id="search-dropdown-root_contactId"
                            placeholder="Select a Contact"
                            spellcheck="false"
                            type="text"
                            value="Mister Test"
                          />
                          <div
                            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                          >
                            <button
                              aria-label="Clear"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                              tabindex="-1"
                              title="Clear"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                                data-testid="CloseIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                            <button
                              aria-label="Open"
                              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                              tabindex="-1"
                              title="Open"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                                data-testid="ArrowDropDownIcon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M7 10l5 5 5-5z"
                                />
                              </svg>
                              <span
                                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                              />
                            </button>
                          </div>
                          <fieldset
                            aria-hidden="true"
                            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                          >
                            <legend
                              class="css-nnbavb"
                            >
                              <span
                                class="notranslate"
                              >
                                â€‹
                              </span>
                            </legend>
                          </fieldset>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="sc-furwcr sc-pVTFL kGJjgX eqAbko pg-grid-col right-aligned-column"
              width="20"
            >
              <button
                class="sc-jcFjpl VmKuo pg-button"
                id="zch6pzpajlpswuvgpls4b"
              >
                Remove
              </button>
            </div>
          </div>
          <div
            class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
          >
            <div
              class="sc-furwcr sc-pVTFL XpgcI eqAbko pg-grid-col"
              width="100"
            >
              <button
                class="sc-jcFjpl dFodJy pg-button"
                id="0ryiwpifbzlblb6c2skhkld"
                style="margin-right: auto;"
              >
                Add
              </button>
            </div>
          </div>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
